### PR TITLE
Bug 1226555 - [Messages] The Conversation split view is not rendered anymore. r=julien

### PR DESCRIPTION
**Pull request by azasypkin**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a0167b8f7f362ad8d0dde0600411f6317d4d554c

File: apps/sms/test/marionette/inbox_test.js [hunk 1]
```diff
@@ -76,4 +76,75 @@ marionette('Inbox View tests', function() {
...
+    });
+
+    test('User can enter New Message view from Inbox and go back', function() {
+      var inboxView = new InboxView(client);
+
+      // Make sure we've entered new message view.
+      var newMessageView = inboxView.createNewMessage();
+      newMessageView.assertRecipientsInputFocused();
+
+      // Workaround for bug 1227147.
```
Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: Just filed a new bug, see https://bugzilla.mozilla.org/show_bug.cgi?id=1227147

File: apps/sms/test/marionette/inbox_test.js [hunk 2]
```diff
@@ -76,4 +76,75 @@ marionette('Inbox View tests', function() {
       });
     });
   });
+
+  suite('Split Inbox view tests', function() {
```
Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: Very basic tests, the last one will fail on master, and passes with the PR.

File: apps/sms/test/marionette/lib/views/inbox/view.js [hunk 3]
```diff
@@ -36,9 +36,11 @@ InboxView.prototype = {
   goToConversation: function(conversationId) {
     var conversation = this.accessors.findConversation(conversationId);
 
+    var isConversationLessDraft = !!conversation.getAttribute('data-draft-id');
```
Comment (azasypkin) [inline apps/sms/test/marionette/lib/views/inbox/view.js]: We change document with the tap, so if we don't get attribute value before we tap - we'll get StaleElementException :/

File: apps/sms/test/marionette/mocks/mock_split_view_mode.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+'use strict';
+
+Services.obs.addObserver(function(document) {
+  if (!document || !document.location ||
+      document.location.origin !== 'app://sms.gaiamobile.org') {
+    return;
+  }
+
+  // Switch to split-view mode Inbox instead.
+  if(document.location.pathname === '/index.html') {
```
Comment (azasypkin) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: Haaaaaaaaack! But will help us to avoid some bit-rotting :) What do you think?

Comment (julienw) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: wondering if we could add hidden entry points to our manifest, or just replace the manifest on the fly before generating the profile and launching mulet... I'm sure this was done in other apps BTW, I think I heard of it. Maybe ask Johan ? But let's not wait for this to move forward, please just file a separate bug.

Comment (julienw) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: nit: please add a space after `if`

File: apps/sms/test/marionette/mocks/mock_split_view_mode.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/* global Services */
```
Comment (julienw) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: nit: maybe add a small comment explaining what this is for ?

File: apps/sms/views/shared/test/unit/navigation_test.js [hunk 6]
```diff
@@ -787,9 +787,11 @@ suite('navigation >', function() {
       function(done) {
         Object.defineProperty(document, 'readyState', { value: 'interactive' });
 
-        // For the split-view case Navigation.init waits for the setReady call
-        // that is made in the suite setup function above.
-        Navigation.init().then(done, done);
+        var NonReadyNavigation = NavigationFactory(fakeWindow);
```
Comment (julienw) [inline apps/sms/views/shared/test/unit/navigation_test.js]: just wondering why this change ?

File: apps/sms/test/marionette/lib/views/inbox/view.js (hunk 3 above)
Comment (julienw) [inline apps/sms/test/marionette/lib/views/inbox/view.js]: nit: please add a comment explaining this :) "in split views mode, etc..."

File: apps/sms/test/marionette/inbox_test.js (hunk 1 above)
Comment (julienw) [inline apps/sms/test/marionette/inbox_test.js]: looks like it may land really soon, you may be able to remove the workaround before landing

File: apps/sms/test/marionette/inbox_test.js [hunk 7]
```diff
@@ -76,4 +76,75 @@ marionette('Inbox View tests', function() {
...
+      );
+
+      messagesApp.launch();
+    });
+
+    test('All conversations are loaded', function() {
+      var inboxView = new InboxView(client);
+
+      client.waitFor(function() {
+        return inboxView.conversations.length === 10;
```
Comment (julienw) [inline apps/sms/test/marionette/inbox_test.js]: nit: please put this magic value in a constant

File: apps/sms/test/marionette/inbox_test.js [hunk 8]
```diff
@@ -76,4 +76,75 @@ marionette('Inbox View tests', function() {
...
+      });
+    });
+
+    test('User can enter Conversation view from Inbox and go back', function() {
+      var inboxView = new InboxView(client);
+
+      // Make sure we've entered conversation view.
+      var conversationView = inboxView.goToConversation(1);
+      assert.equal(conversationView.messages().length, 5);
```
Comment (julienw) [inline apps/sms/test/marionette/inbox_test.js]: please have a constant for this magic value as well :)

File: apps/sms/views/shared/test/unit/navigation_test.js (hunk 6 above)
Comment (azasypkin) [inline apps/sms/views/shared/test/unit/navigation_test.js]: Because ```setReady``` is called in setup so that other tests within this suite work as expected. I had two choices to make it fail on master, either this, or move it to another suite - chose the simplest - what do you prefer?

File: apps/sms/test/marionette/inbox_test.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: Yeah, I'm planning to verify with their PR before landing and remove workaround if it helps.

File: apps/sms/test/marionette/mocks/mock_split_view_mode.js (hunk 5 above)
Comment (azasypkin) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: Ah, yeah, sure!

File: apps/sms/views/shared/test/unit/navigation_test.js (hunk 6 above)
Comment (julienw) [inline apps/sms/views/shared/test/unit/navigation_test.js]: got it, it's good like this :)

Comment (julienw) [inline apps/sms/views/shared/test/unit/navigation_test.js]: maybe change the test title then ?

File: apps/sms/test/marionette/inbox_test.js (hunk 7 above)
Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: heh, yeah, thanks :)

File: apps/sms/test/marionette/inbox_test.js (hunk 8 above)
Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: Okay :)

File: apps/sms/test/marionette/mocks/mock_split_view_mode.js (hunk 4 above)
Comment (azasypkin) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: > Maybe ask Johan ? 

Yeah, will ask him, thanks! :)

> just replace the manifest on the fly before generating the profile and launching mulet...

I believe it will be a pain since we want to run tests for both versions on the same profile....

> if we could add hidden entry points to our manifest,

Hmm interesting, depending on how it will look like. But generally I'd rather make hacky test than hack real manifest :p

File: apps/sms/views/shared/test/unit/navigation_test.js (hunk 6 above)
Comment (azasypkin) [inline apps/sms/views/shared/test/unit/navigation_test.js]: Yep!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dd0e4a398748f05fb14c128134245fc93b6d8b2c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a22d0d281e165098326ee54bd15cccc938e371ff

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e25e984658995e282dbcaddd7ece167a555aeb94

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b7a22b1bba5269b28d3ea6ed79885260dedcf2ca

File: apps/sms/test/marionette/mocks/mock_split_view_mode.js (hunk 4 above)
Comment (JohanLorenzo) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: IMO, this hack looks simpler and nicer than a manifest injection :)

I couldn't find it in the w3c specs: `location.replace()` is a synchronous function, right?

Comment (azasypkin) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: Mmm, what do you mean by synchronous function? If you mean that here we guarantee that application will not start before URL is changed - then I believe yes - we call ```replace```, that will fire all ```document-element-inserted``` observers once again before app is actually started (including this one).

Comment (julienw) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: this triggers a full navigation, but "replace" will also replace the current history entry (as opposed to .assign or setting href directly): https://developer.mozilla.org/en-US/docs/Web/API/Location/replace

because of the navigation eveything in the page is stopped and discarded anyway (JS/CSS loading, and even all execution).

Comment (julienw) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: still "nit: please add a space after if" ;)

Comment (azasypkin) [inline apps/sms/test/marionette/mocks/mock_split_view_mode.js]: grr sorry have missed that :/

File: apps/sms/test/marionette/inbox_test.js [hunk 9]
```diff
@@ -76,4 +76,69 @@ marionette('Inbox View tests', function() {
       });
     });
   });
+
+  suite('Split Inbox view tests', function() {
+    var numberOfConversations = 10;
+    var numberOfMessagesInConversation = 5;
```
Comment (julienw) [inline apps/sms/test/marionette/inbox_test.js]: nit: I'd use capital letters, to better indicate they're constants. Don't know what you think ? Also maybe we can use "const" now that we use node v4 ?

Comment (azasypkin) [inline apps/sms/test/marionette/inbox_test.js]: Ah didn't think that const works in tests now! And yeah with ```const``` all-caps will be better

**Comment by julienw** [comment]
Now I see this error when coming back from a Conversation to Inbox: `TypeError: this.recipients is null`. I added some log, and here is the stack:
```
updateDraft@app://sms.gaiamobile.org/views/conversation/js/conversation.js:2982:17
conv_onVisibilityChange@app://sms.gaiamobile.org/views/conversation/js/conversation.js:326:7
EventListener.handleEvent*conv_init@app://sms.gaiamobile.org/views/conversation/js/conversation.js:148:1
init@app://sms.gaiamobile.org/views/conversation/js/startup.js:74:7
@app://sms.gaiamobile.org/views/shared/js/bootstrap.js:6:1
@app://sms.gaiamobile.org/views/shared/js/bootstrap.js:3:1
```

So that means we get a visibilitychange event when we navigate ? looks weird... But I think we should handle this in a separate bug. Would you file it ?

**Comment by azasypkin** [comment]
>  So that means we get a visibilitychange event when we navigate ? looks weird... But I think we should handle this in a separate bug. Would you file it ?

Yeah, it's old one that I wanted to file bug for some time already :) It's ```updateDraft``` called in visibility change handler, but this.activeThread has been nullified already so that ConversationView thinks we're in New Message view.

**Comment by julienw** [comment]
yeah but I don't get why visibilitychange is getting sent in that case; I'll look in the spec :)

**Comment by azasypkin** [comment]
Filed: https://bugzilla.mozilla.org/show_bug.cgi?id=1229386

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c6769b763aaf8c5ee4c95e9f2a496072117edf19
